File: src/games/sudoku/gameplay/analysis/techniqueModules/hiddenSingle.ts
import { sudokuDigits } from '../../../types';
import { allHouseRefs, buildPlacementMove, getHouseDigitMatches } from '../techniqueHelpers';
import type { SudokuTechniqueDispatcher } from '../techniqueModuleTypes';

const HOUSE_SCAN_WEIGHTS = { row: 1.2, column: 1.4, box: 1.0 } as const;

export const hiddenSingleTechnique: SudokuTechniqueDispatcher = {
  technique: 'hidden-single',
  tier: 'easy',
  findMove(state) {
    let best = null as ReturnType<typeof buildPlacementMove> | null;
    let bestComplexity = Infinity;

    for (const { house, cells } of allHouseRefs) {
      const emptyCells = cells.filter((i) => state.board[i] === 0).length;
      const houseComplexity = emptyCells * HOUSE_SCAN_WEIGHTS[house.kind];
      if (houseComplexity >= bestComplexity) {
        continue;
      }

      for (const digit of sudokuDigits) {
        const matches = getHouseDigitMatches(state, cells, digit);
        if (matches.length !== 1) {
          continue;
        }

        bestComplexity = houseComplexity;
        const targetIndex = matches[0];
        const row = Math.floor(targetIndex / 9);
        const col = targetIndex % 9;
        best = buildPlacementMove({
          technique: 'hidden-single',
          row,
          col,
          digit,
          evidenceCells: cells,
          houses: [
            house,
            ...[
              { kind: 'row' as const, index: row },
              { kind: 'column' as const, index: col },
              { kind: 'box' as const, index: Math.floor(row / 3) * 3 + Math.floor(col / 3) },
            ].filter((ref) => !(ref.kind === house.kind && ref.index === house.index)),
          ],
          complexity: houseComplexity,
        });
        break; // one match per house is enough — all digits in this house share the same complexity
      }
    }

    return best;
  },
};

File: engine/src/main/kotlin/com/quietgrid/engine/sudoku/techniques/HiddenSingle.kt
package com.quietgrid.engine.sudoku.techniques

import com.quietgrid.engine.core.Difficulty
import com.quietgrid.engine.sudoku.*

private val HOUSE_SCAN_WEIGHTS = mapOf("row" to 1.2, "column" to 1.4, "box" to 1.0)

object HiddenSingleTechnique : SudokuTechniqueDispatcher {
    override val technique = SudokuTechnique.HIDDEN_SINGLE
    override val tier = Difficulty.EASY

    override fun findMove(state: SudokuBitmaskState): SudokuCanonicalMove? {
        var best: SudokuPlacementMove? = null
        var bestComplexity = Double.MAX_VALUE

        for (ref in allHouseRefs) {
            val emptyCells = ref.cells.count { state.board[it] == 0 }
            val houseComplexity = emptyCells * HOUSE_SCAN_WEIGHTS.getValue(ref.house.kind)
            if (houseComplexity >= bestComplexity) continue

            for (digit in 1..9) {
                val matches = getHouseDigitMatches(state, ref.cells, digit)
                if (matches.size != 1) continue

                bestComplexity = houseComplexity
                val targetIndex = matches[0]
                val row = targetIndex / 9
                val col = targetIndex % 9
                val extraHouses = listOf(
                    SudokuHouseRef("row", row),
                    SudokuHouseRef("column", col),
                    SudokuHouseRef("box", row / 3 * 3 + col / 3),
                ).filter { !(it.kind == ref.house.kind && it.index == ref.house.index) }
                best = buildPlacementMove(
                    technique = SudokuTechnique.HIDDEN_SINGLE,
                    row = row, col = col, digit = digit,
                    evidenceCells = ref.cells,
                    houses = listOf(ref.house) + extraHouses,
                    complexity = houseComplexity.toInt(),
                )
                break
            }
        }
        return best
    }
}
